salt/pillar/__init__.py

```diff
@@ -260,40 +272,24 @@ class Pillar(object):
     '''
     def __init__(self, opts, grains, minion_id, saltenv, ext=None, functions=None,
                  pillar=None, pillarenv=None):
+
+        self.orig_opts = opts
+        self._functions = functions
         self.minion_id = minion_id
         # Store the file_roots path so we can restore later. Issue 5449
-        self.actual_file_roots = opts['file_roots']
+        self.actual_file_roots = self.orig_opts['file_roots']
         # use the local file client
-        self.opts = self.__gen_opts(opts, grains, saltenv=saltenv, ext=ext, pillarenv=pillarenv)
-        self.client = salt.fileclient.get_file_client(self.opts, True)
+        self._opts = self.__gen_opts(self.orig_opts, grains, saltenv=saltenv, ext=ext, pillarenv=pillarenv)
 
-        if opts.get('file_client', '') == 'local':
-            opts['grains'] = grains
+        # FIXME: if we change self.orig_opts then it is no longer "original".  Hmmm.
+        if self.orig_opts.get('file_client') == 'local':
+            self.orig_opts['grains'] = grains
 
-        # if we didn't pass in functions, lets load them
-        if functions is None:
-            utils = salt.loader.utils(opts)
-            if opts.get('file_client', '') == 'local':
-                self.functions = salt.loader.minion_mods(opts, utils=utils)
-            else:
-                self.functions = salt.loader.minion_mods(self.opts, utils=utils)
-        else:
-            self.functions = functions
+        self.merge_strategy = opts.get('pillar_source_merging_strategy', 'smart')
 
-        self.matcher = salt.minion.Matcher(self.opts, self.functions)
-        self.rend = salt.loader.render(self.opts, self.functions)
-        ext_pillar_opts = copy.deepcopy(self.opts)
-        # Fix self.opts['file_roots'] so that ext_pillars know the real
-        # location of file_roots. Issue 5951
-        ext_pillar_opts['file_roots'] = self.actual_file_roots
-        # Keep the incoming opts ID intact, ie, the master id
-        if 'id' in opts:
-            ext_pillar_opts['id'] = opts['id']
-        self.merge_strategy = 'smart'
-        if opts.get('pillar_source_merging_strategy'):
-            self.merge_strategy = opts['pillar_source_merging_strategy']
+        self._pillar_accumulator = {}
```

Comment:
``self._pillar_accumulator`` is the state for the ``Pillar`` object and accumulates each "layer" of pillar that is generated.